File: enchiridion-go/internal/sessionstate/sessionstate_test.go
package sessionstate

import (
	"os"
	"path/filepath"
	"testing"
)

func env(m map[string]string) func(string) (string, bool) {
	return func(key string) (string, bool) {
		v, ok := m[key]
		return v, ok
	}
}

func TestWriteThenReadRoundTrips(t *testing.T) {
	dir := t.TempDir()
	if err := WriteTranscriptPath("session-a", "/path/to/a.jsonl", dir); err != nil {
		t.Fatal(err)
	}
	got, ok := ReadTranscriptPath("session-a", dir)
	if !ok || got != "/path/to/a.jsonl" {
		t.Fatalf("got (%q, %v), want (/path/to/a.jsonl, true)", got, ok)
	}
}

func TestReadReturnsFalseForUnknownSession(t *testing.T) {
	if _, ok := ReadTranscriptPath("never-written", t.TempDir()); ok {
		t.Fatal("unknown session reported as present")
	}
}

func TestSessionsAreIsolated(t *testing.T) {
	dir := t.TempDir()
	if err := WriteTranscriptPath("session-a", "/path/to/a.jsonl", dir); err != nil {
		t.Fatal(err)
	}
	if err := WriteTranscriptPath("session-b", "/path/to/b.jsonl", dir); err != nil {
		t.Fatal(err)
	}
	if got, _ := ReadTranscriptPath("session-a", dir); got != "/path/to/a.jsonl" {
		t.Errorf("session-a = %q", got)
	}
	if got, _ := ReadTranscriptPath("session-b", dir); got != "/path/to/b.jsonl" {
		t.Errorf("session-b = %q", got)
	}
}

func TestWriteOverwritesPreviousValue(t *testing.T) {
	dir := t.TempDir()
	if err := WriteTranscriptPath("session-a", "/old.jsonl", dir); err != nil {
		t.Fatal(err)
	}
	if err := WriteTranscriptPath("session-a", "/new.jsonl", dir); err != nil {
		t.Fatal(err)
	}
	if got, _ := ReadTranscriptPath("session-a", dir); got != "/new.jsonl" {
		t.Errorf("got %q, want /new.jsonl", got)
	}
}

func TestReadReturnsFalseForCorruptedState(t *testing.T) {
	dir := t.TempDir()
	if err := os.WriteFile(filepath.Join(dir, "session-a.json"), []byte("not json"), 0o644); err != nil {
		t.Fatal(err)
	}
	if _, ok := ReadTranscriptPath("session-a", dir); ok {
		t.Fatal("corrupted state reported as present")
	}
}

func TestWriteCreatesStateDirIfMissing(t *testing.T) {
	nested := filepath.Join(t.TempDir(), "nested", "sessions")
	if err := WriteTranscriptPath("session-a", "/a.jsonl", nested); err != nil {
		t.Fatal(err)
	}
	if got, _ := ReadTranscriptPath("session-a", nested); got != "/a.jsonl" {
		t.Errorf("got %q", got)
	}
}

func TestSessionsDirUnderGivenRoot(t *testing.T) {
	got := SessionsDir("/some/project", "", env(nil))
	want := filepath.Join("/some/project", ".claude", "wiki-knowledge", "sessions")
	if got != want {
		t.Errorf("got %q, want %q", got, want)
	}
}

func TestSessionsDirDefaultsToCwd(t *testing.T) {
	dir := t.TempDir()
	got := SessionsDir("", dir, env(nil))
	want := filepath.Join(dir, ".claude", "wiki-knowledge", "sessions")
	if got != want {
		t.Errorf("got %q, want %q", got, want)
	}
}

func TestSessionsDirWalksUpForClaudeAncestor(t *testing.T) {
	root := t.TempDir()
	if err := os.MkdirAll(filepath.Join(root, ".claude"), 0o755); err != nil {
		t.Fatal(err)
	}
	subdir := filepath.Join(root, "src", "deep")
	if err := os.MkdirAll(subdir, 0o755); err != nil {
		t.Fatal(err)
	}
	got := SessionsDir("", subdir, env(nil))
	want := filepath.Join(root, ".claude", "wiki-knowledge", "sessions")
	if got != want {
		t.Errorf("got %q, want %q", got, want)
	}
}

func TestSessionsDirProjectDirOverridesWalkup(t *testing.T) {
	stateRoot := filepath.Join(t.TempDir(), "elsewhere")
	if err := os.MkdirAll(stateRoot, 0o755); err != nil {
		t.Fatal(err)
	}
	cwd := t.TempDir()
	got := SessionsDir("", cwd, env(map[string]string{"CLAUDE_PROJECT_DIR": stateRoot}))
	want := filepath.Join(stateRoot, ".claude", "wiki-knowledge", "sessions")
	if got != want {
		t.Errorf("got %q, want %q", got, want)
	}
}

func TestSessionsDirRootInjectionSkipsEnvAndCwd(t *testing.T) {
	got := SessionsDir("/explicit", t.TempDir(), env(map[string]string{"CLAUDE_PROJECT_DIR": "/should/be/ignored"}))
	want := filepath.Join("/explicit", ".claude", "wiki-knowledge", "sessions")
	if got != want {
		t.Errorf("got %q, want %q", got, want)
	}
}

File: enchiridion-ts/src/sessionstate.test.ts
/**
 * Unit tests for the sessionstate module (#257). Uses mock env lookups and
 * temp-dir state so no real Claude Code environment or hook is touched.
 */

import { test } from "node:test";
import assert from "node:assert/strict";
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import {
  sessionsDir,
  writeTranscriptPath,
  readTranscriptPath,
} from "./sessionstate.js";
import type { LookupEnv } from "./sessionstate.js";

function tmp(): string {
  return fs.mkdtempSync(path.join(os.tmpdir(), "sessionstate-test-"));
}

/** Build a LookupEnv from a fixed map (missing keys report not-present). */
function env(map: Record<string, string>): LookupEnv {
  return (key: string): [string | undefined, boolean] => {
    const value = map[key];
    return [value, value !== undefined];
  };
}

// ---------------------------------------------------------------------------
// sessionsDir resolution order
// ---------------------------------------------------------------------------

test("sessionsDir: injected root wins over everything", () => {
  const root = tmp();
  const got = sessionsDir(root, "/some/other/cwd", env({}));
  assert.equal(got, path.join(root, ".claude", "wiki-knowledge", "sessions"));
});

test("sessionsDir: CLAUDE_PROJECT_DIR is used when no root", () => {
  const project = tmp();
  const got = sessionsDir(
    "",
    "/some/cwd",
    env({ CLAUDE_PROJECT_DIR: project }),
  );
  assert.equal(
    got,
    path.join(project, ".claude", "wiki-knowledge", "sessions"),
  );
});

test("sessionsDir: walks up to the nearest .claude ancestor", () => {
  const root = tmp();
  const nested = path.join(root, "a", "b");
  fs.mkdirSync(path.join(root, ".claude"), { recursive: true });
  const got = sessionsDir("", nested, env({}));
  assert.equal(got, path.join(root, ".claude", "wiki-knowledge", "sessions"));
});

test("sessionsDir: falls back to cwd when no .claude ancestor exists", () => {
  const cwd = tmp();
  const got = sessionsDir("", cwd, env({}));
  assert.equal(got, path.join(cwd, ".claude", "wiki-knowledge", "sessions"));
});

test("sessionsDir: empty CLAUDE_PROJECT_DIR is ignored", () => {
  const cwd = tmp();
  const got = sessionsDir("", cwd, env({ CLAUDE_PROJECT_DIR: "" }));
  assert.equal(got, path.join(cwd, ".claude", "wiki-knowledge", "sessions"));
});

// ---------------------------------------------------------------------------
// write / read transcript path
// ---------------------------------------------------------------------------

test("writeTranscriptPath then readTranscriptPath round-trips", () => {
  const stateDir = path.join(tmp(), ".claude", "wiki-knowledge", "sessions");
  writeTranscriptPath("sess-123", "/tmp/transcript.jsonl", stateDir);
  assert.equal(
    readTranscriptPath("sess-123", stateDir),
    "/tmp/transcript.jsonl",
  );
});

test("writeTranscriptPath creates the state directory", () => {
  const stateDir = path.join(tmp(), "nested", "state");
  writeTranscriptPath("sess-1", "/t.jsonl", stateDir);
  assert.ok(fs.statSync(stateDir).isDirectory());
  const raw = fs.readFileSync(path.join(stateDir, "sess-1.json"), "utf8");
  assert.deepEqual(JSON.parse(raw), { transcript_path: "/t.jsonl" });
});

test("readTranscriptPath is undefined for a missing session", () => {
  const stateDir = path.join(tmp(), "state");
  assert.equal(readTranscriptPath("nope", stateDir), undefined);
});

test("readTranscriptPath is undefined for unparsable state", () => {
  const stateDir = path.join(tmp(), "state");
  fs.mkdirSync(stateDir, { recursive: true });
  fs.writeFileSync(path.join(stateDir, "bad.json"), "not json");
  assert.equal(readTranscriptPath("bad", stateDir), undefined);
});

test("readTranscriptPath is undefined when transcript_path is not a string", () => {
  const stateDir = path.join(tmp(), "state");
  fs.mkdirSync(stateDir, { recursive: true });
  fs.writeFileSync(
    path.join(stateDir, "x.json"),
    JSON.stringify({ transcript_path: 42 }),
  );
  assert.equal(readTranscriptPath("x", stateDir), undefined);
});
